packages/reporting/src/request/utils/chrome-storage-map.js

```diff
@@ -0,0 +1,240 @@
+    hardFlushIntervalInMs = 1000,
+    ttlInMs = 7 * 24 * 60 * 60 * 1000 /* 1 week */,
+    maxEntries = 5000,
+  }) {
+    if (!storageKey) {
+      throw new Error('Missing storage key');
+    }
+    this.storageKey = storageKey;
+    this.inMemoryMap = new Map();
+    this._initialSyncComplete = false;
+    this.maxEntries = maxEntries;
+
+    // Make sure old entries that were not cleaned up are eventually
+    // removed. Otherwise, we could exceed the local storage quota.
+    // Plus, when the maps get big, serializing and deserializing
+    // may become expensive. If the actively triggered clean up works,
+    // there should be no need to make this expiration too aggressive.
+    this.ttlInMs = ttlInMs;
+    this._ttlMap = new Map();
+
+    // Flush handling logic: the difference between both limits is that
+    // the soft limit does not guarantee that a flush will eventually
+    // be performed. After each write operation, it will reset the soft
+    // timeout and then flush. Thus, if you keep writing, it will never
+    // flush. The hard limit, on the other hand, forces that data gets
+    // persisted, but could result in ill-timed write operations.
+    //
+    // If there are bursts of operations, ideally you want to flush
+    // at the end of the burst. The soft limit will result in that,
+    // while the hard limit mitigates the risk that the script
+    // gets killed before the data gets persisted.
+    //
+    // Rule of thumbs:
+    // * The soft limit should be lower then the hard limit
+    // * The hard limit should not be set too high. Remember, it is
+    //   the protection against the browser unpredictably killing
+    //   the execution.
+    this.softFlushIntervalInMs = softFlushIntervalInMs;
+    this.hardFlushIntervalInMs = hardFlushIntervalInMs;
+    this._scheduledFlush = null;
+    this._lastFlush = Date.now();
+    this._dirty = false;
+
+    // Assumption: there should be enough time during startup to load
+    // the persisted map. Otherwise, the state will be inconsistent
+    // whenever the script is loaded (it will eventually become consistent,
+    // but that will not help if the browser kills it quickly).
+    //
+    // (If that assumption does not hold, _warnIfOutOfSync will detect
+    // and log it. A potential improvement could be to treat the
+    // in-memory map as the source of truth in that scenario.)
+    this._pending = new Promise((resolve, reject) => {
+      chrome.storage.local.get([this.storageKey], (result) => {
+        if (chrome.runtime.lastError) {
+          reject(chrome.runtime.lastError);
+        } else {
+          const { entries = {}, ttl = {} } = result[this.storageKey] || {};
+          this.inMemoryMap = new Map(Object.entries(entries));
+          this._ttlMap = new Map(Object.entries(ttl));
+          this._initialSyncComplete = true;
+          this.expireOldEntries();
+          resolve();
+        }
+      });
+    });
+    setInterval(() => {
```

Comment:
required for short lived values

Reply:
I don't think it is a good idea to have something like that if it is indeed to be a general data structure. First, it is in the constructor and never cleans up. The other is that it will not play nicely if someone uses it with short TTLs.

Reply:
Automatically cleaning up is not trivial. AutoSyncingMap was more like Redis that does it lazily. I feel that the alternative of letting the user of such data occasionally force a full scan (if necessary) is the better approach. As a compromise, it performed a scan once on startup, but startup is already O(n), so here the operation is not adding costs (at least, not from complexity).